Mathlib/Algebra/Order/Ring/StandardPart.lean

```diff
@@ -0,0 +1,323 @@
+public import Mathlib.Algebra.Ring.Subring.Order
+public import Mathlib.Data.Real.Archimedean
+public import Mathlib.Order.Quotient
+public import Mathlib.RingTheory.Valuation.ValuationSubring
+
+import Mathlib.Data.Real.CompleteField
+
+/-!
+# Standard part function
+
+Given a finite element in a non-archimedean field, the standard part function rounds it to the
+unique closest real number. That is, it chops off any infinitesimals.
+
+Let `K` be a linearly ordered field. The subset of finite elements (i.e. those bounded by a natural
+number) is a `ValuationSubring`, which means we can construct its residue field
+`FiniteResidueField`, roughly corresponding to the finite elements quotiented by infinitesimals.
+This field inherits a `LinearOrder` instance, which makes it into an Archimedean linearly ordered
+field, meaning we can uniquely embed it in the reals.
+
+Given a finite element of the field, the `ArchimedeanClass.standardPart` function returns the real
+number corresponding to this unique embedding. This function generalizes, among other things, the
+standard part function on `Hyperreal`.
+
+## TODO
+
+Redefine `Hyperreal.st` in terms of `ArchimedeanClass.standardPart`.
+
+## References
+
+* https://en.wikipedia.org/wiki/Standard_part_function
+-/
+
+@[expose] public section
+
+namespace ArchimedeanClass
+variable
+  {K : Type*} [LinearOrder K] [Field K] [IsOrderedRing K] {x y : K}
+  {R : Type*} [LinearOrder R] [CommRing R] [IsOrderedRing R] [Archimedean R]
+
+/-! ### Finite residue field -/
+
+variable (K) in
+/-- The valuation subring of elements in non-negative Archimedean classes, i.e. elements bounded by
+some natural number. -/
+protected noncomputable def Finite : Type _ :=
+  (addValuation K).toValuation.valuationSubring
+
+instance : CommRing (ArchimedeanClass.Finite K) := by
+  unfold ArchimedeanClass.Finite; infer_instance
+
+instance : IsDomain (ArchimedeanClass.Finite K) := by
+  unfold ArchimedeanClass.Finite; infer_instance
+
+instance : ValuationRing (ArchimedeanClass.Finite K) := by
+  unfold ArchimedeanClass.Finite; infer_instance
+
+instance : LinearOrder (ArchimedeanClass.Finite K) := by
+  unfold ArchimedeanClass.Finite; infer_instance
+
+instance : IsStrictOrderedRing (ArchimedeanClass.Finite K) := by
+  unfold ArchimedeanClass.Finite; infer_instance
+
+namespace Finite
+
+@[simp] theorem val_zero : (0 : ArchimedeanClass.Finite K).1 = 0 := rfl
+@[simp] theorem val_one : (1 : ArchimedeanClass.Finite K).1 = 1 := rfl
+@[simp] theorem val_add (x y : ArchimedeanClass.Finite K) : (x + y).1 = x.1 + y.1 := rfl
+@[simp] theorem val_sub (x y : ArchimedeanClass.Finite K) : (x - y).1 = x.1 - y.1 := rfl
+@[simp] theorem val_mul (x y : ArchimedeanClass.Finite K) : (x * y).1 = x.1 * y.1 := rfl
+
+@[ext] theorem ext {x y : ArchimedeanClass.Finite K} (h : x.1 = y.1) : x = y := Subtype.ext h
+
+/-- The constructor for `ArchimedeanClass.Finite`. -/
+protected def mk (x : K) (h : 0 ≤ mk x) : ArchimedeanClass.Finite K := ⟨x, h⟩
```

Comment:
Why `protected`?

Reply:
I often want to refer to `Finite.mk` and `ArchimedeanClass.mk` in the same theorem.